firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
+
+
+def factor_interior_mat(A00):
+    # Assume that interior DOF list i0 is ordered such that A00 is block diagonal
+    # with blocks of increasing dimension
+    indptr, indices, data = A00.getValuesCSR()
+    degree = numpy.diff(indptr)
+
+    # TODO handle non-symmetric case with LU, requires scipy
+    invchol = lambda X: numpy.linalg.inv(numpy.linalg.cholesky(X))
+    nblocks = numpy.count_nonzero(degree == 1)
+    zlice = slice(0, nblocks)
+    numpy.sqrt(data[zlice], out=data[zlice])
+    numpy.reciprocal(data[zlice], out=data[zlice])
+    PETSc.Log.logFlops(2*nblocks)
+    for k in range(2, degree[-1]+1):
+        nblocks = numpy.count_nonzero(degree == k)
+        zlice = slice(zlice.stop, zlice.stop + k*nblocks)
+        data[zlice] = invchol(data[zlice].reshape((-1, k, k))).reshape((-1,))
+        flops = ((k+1)**3 + 5*(k+1)-12)//3 + k**3
+        PETSc.Log.logFlops(flops*nblocks)
+
+    A00.setValuesCSR(indptr, indices, data)
+    A00.assemble()
+
+
+@PETSc.Log.EventDecorator("FDMCondense")
+def condense_element_mat(A, i0, i1, submats):
+    isrows = [i0, i0, i1, i1]
+    iscols = [i0, i1, i0, i1]
+    submats[:4] = A.createSubMatrices(isrows, iscols=iscols, submats=submats[:4] if submats[0] else None)
+    A00, A01, A10, A11 = submats[:4]
+    factor_interior_mat(A00)
+    submats[4] = A00.matMult(A01, result=submats[4])
+    submats[5] = A10.matTransposeMult(A00, result=submats[5])
+    submats[6] = submats[5].matMult(submats[4], result=submats[6])
+    submats[6].aypx(-1.0, A11)
+    return submats[6]
+
+
+@PETSc.Log.EventDecorator("FDMCondense")
+def condense_element_pattern(A, i0, i1, submats):
+    isrows = [i0, i0, i1]
+    iscols = [i0, i1, i0]
+    submats[:3] = A.createSubMatrices(isrows, iscols=iscols, submats=submats[:3] if submats[0] else None)
+    A00, A01, A10 = submats[:3]
+    submats[4] = A10.matTransposeMult(A00, result=submats[4])
+    submats[5] = A00.matMult(A01, result=submats[5])
+    submats[6] = submats[4].matMult(submats[5], result=submats[6])
+    submats[6].aypx(0.0, A)
+    return submats[6]
+
+
+@PETSc.Log.EventDecorator("LoadCode")
+def load_c_code(code, name, **kwargs):
+    from pyop2.compilation import load
+    from pyop2.utils import get_petsc_dir
+    cppargs = ["-I%s/include" % d for d in get_petsc_dir()]
+    ldargs = (["-L%s/lib" % d for d in get_petsc_dir()]
+              + ["-Wl,-rpath,%s/lib" % d for d in get_petsc_dir()]
+              + ["-lpetsc", "-lm"])
+    funptr = load(code, "c", name,
+                  cppargs=cppargs, ldargs=ldargs,
+                  **kwargs)
+
+    def get_pointer(obj):
+        if isinstance(obj, (PETSc.Mat, PETSc.Vec)):
+            return obj.handle
+        elif isinstance(obj, numpy.ndarray):
+            return obj.ctypes.data
+        return obj
+
+    @PETSc.Log.EventDecorator(name)
+    def wrapper(*args):
+        return funptr(*list(map(get_pointer, args)))
+    return wrapper
+
+
+def load_assemble_csr(comm, triu=False):
```

Comment:
Comment "Insert one sparse matrix into another sparse matrix. Done in C for efficiency, since it loops over rows."